File: Ark.java
///usr/bin/env jbang "$0" "$@" ; exit $?
// Usage: java Ark.java <filePaths> <backupDir>
// Usage with JBang: jbang Ark.java <filePaths> <backupDir>
// Example: java Ark.java /path/to/xy.db /opt/backup
// Example: java Ark.java /path/to/xy.db,/path/to/xy.sql /opt/backup
// With debug: DEBUG=1 java Ark.java /path/to/xy.db /opt/backup

import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.temporal.TemporalAdjusters;
import java.util.*;
import java.util.stream.Collectors;
import java.util.zip.*;

public class Ark {
    private static final boolean DEBUG = "1".equals(System.getenv("DEBUG"));
    private static final DateTimeFormatter FMT = DateTimeFormatter.ofPattern("yyyyMMdd");

    private static void log(Object... args) {
        if (DEBUG) System.out.println(Arrays.stream(args).map(String::valueOf).collect(Collectors.joining(" ")));
    }

    private static void createBackup(List<String> filePaths, String backupDir) throws IOException {
        LocalDate now = LocalDate.now();
        String filename = "export" + now.format(FMT) + ".zip";
        Path outputPath = Paths.get(backupDir, filename);

        if (Files.exists(outputPath)) {
            System.out.println("Backup " + filename + " already exists. Skipping creation.");
            return;
        }

        long start = System.nanoTime();
        try (ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(outputPath.toFile()))) {
            for (String filePath : filePaths) {
                Path path = Paths.get(filePath);
                ZipEntry entry = new ZipEntry(path.getFileName().toString());
                zos.putNextEntry(entry);
                Files.copy(path, zos);
                zos.closeEntry();
            }
        }
        double elapsed = (System.nanoTime() - start) / 1_000_000_000.0;
        System.out.printf("Created backup: %s (%.3fs)%n", filename, elapsed);
    }

    private static void rotateBackups(String backupDir) throws IOException {
        log("[DEBUG] Starting rotateBackups");
        LocalDate now = LocalDate.now();
        List<Path> files = Files.list(Paths.get(backupDir))
            .filter(p -> p.getFileName().toString().startsWith("export") && p.getFileName().toString().endsWith(".zip"))
            .collect(Collectors.toList());
        log("[DEBUG] Backup files found:", files.size());

        Set<String> toKeep = new HashSet<>();

        // Daily: last 7 days
        log("[DEBUG] Processing daily backups");
        for (int i = 0; i < 7; i++) {
            toKeep.add("export" + now.minusDays(i).format(FMT) + ".zip");
        }

        // Weekly: Sundays in current month
        log("[DEBUG] Processing weekly backups");
        LocalDate weekEnd = now.minusDays(7).with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY));
        LocalDate monthStart = now.withDayOfMonth(1);
        int weekCount = 0;
        while (weekEnd.getMonth() == now.getMonth() && !weekEnd.isBefore(monthStart) && weekCount < 10) {
            toKeep.add("export" + weekEnd.format(FMT) + ".zip");
            weekEnd = weekEnd.minusDays(7).with(TemporalAdjusters.nextOrSame(DayOfWeek.SUNDAY));
            weekCount++;
        }

        // Monthly: last day of previous months in current year
        log("[DEBUG] Processing monthly backups");
        for (int m = 1; m < now.getMonthValue(); m++) {
            LocalDate monthEnd = now.minusMonths(m).with(TemporalAdjusters.lastDayOfMonth());
            toKeep.add("export" + monthEnd.format(FMT) + ".zip");
        }

        // Yearly: Dec 31 of previous years
        log("[DEBUG] Processing yearly backups");
        if (!files.isEmpty()) {
            int minYear = files.stream()
                .map(p -> p.getFileName().toString().substring(6, 10))
                .mapToInt(Integer::parseInt)
                .min().orElse(now.getYear());
            for (int y = minYear; y < now.getYear(); y++) {
                LocalDate yearEnd = LocalDate.of(y, 12, 31);
                toKeep.add("export" + yearEnd.format(FMT) + ".zip");
            }
        }

        // Delete files not in toKeep
        log("[DEBUG] Processing deletions");
        for (Path file : files) {
            String filename = file.getFileName().toString();
            if (!toKeep.contains(filename)) {
                Files.delete(file);
                System.out.println("Deleted old backup: " + filename);
            }
        }
        log("[DEBUG] rotateBackups completed");
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: java Ark.java <filePath1[,filePath2,...]> <backupDir>");
            System.err.println("Example: java Ark.java /path/to/xy.db /opt/backup");
            System.err.println("Example: java Ark.java /path/to/xy.db,/path/to/xy.sql /opt/backup");
            System.exit(1);
        }

        List<String> filePaths = Arrays.stream(args[0].split(","))
            .map(String::trim)
            .collect(Collectors.toList());
        String backupDir = args[1];

        try {
            Files.createDirectories(Paths.get(backupDir));
            createBackup(filePaths, backupDir);
            rotateBackups(backupDir);
            System.out.println("Backup and rotation completed.");
            System.exit(0);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: ark.js
// Script for database backup and rotation
// Usage: node ark.js <filePaths> <backupDir>
// Example: node ark.js /path/to/xy.db /opt/backup
// Example: node ark.js /path/to/xy.db,/path/to/xy.sql /opt/backup
// Dependencies: npm install adm-zip
// bun build --compile --target=bun-linux-x64 ./ark.js --outfile ark
// Schedule with cron: 55 23 * * * /usr/bin/node /path/to/ark.js /path/to/xy.db,/path/to/xy.sql /opt/backup

const fs = require('fs/promises');
const path = require('path');
const AdmZip = require('adm-zip');

const DEBUG = process.env.DEBUG === '1';
const log = (...args) => DEBUG && console.log(...args);

const format = (date) => {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}${m}${d}`;
};

const parse = (str) => {
  const y = parseInt(str.slice(0, 4));
  const m = parseInt(str.slice(4, 6)) - 1;
  const d = parseInt(str.slice(6, 8));
  return new Date(y, m, d);
};

const subDays = (date, days) => new Date(date.getTime() - days * 86400000);
const startOfMonth = (date) => new Date(date.getFullYear(), date.getMonth(), 1);
const endOfMonth = (date) => new Date(date.getFullYear(), date.getMonth() + 1, 0);
const endOfYear = (date) => new Date(date.getFullYear(), 11, 31);
const isSameMonth = (d1, d2) => d1.getFullYear() === d2.getFullYear() && d1.getMonth() === d2.getMonth();
const endOfWeek = (date) => {
  const day = date.getDay();
  const diff = day === 0 ? 0 : 7 - day;
  return new Date(date.getTime() + diff * 86400000);
};

async function createBackup(filePaths, backupDir) {
  const now = new Date();
  const filename = `export${format(now)}.zip`;
  const outputPath = path.join(backupDir, filename);

  // Check if today's backup already exists; skip if it does
  try {
    await fs.access(outputPath);
    console.log(`Backup ${filename} already exists. Skipping creation.`);
    return;
  } catch {
    // Proceed to create
  }

  const startTime = performance.now();
  const zip = new AdmZip();
  filePaths.forEach(filePath => zip.addLocalFile(filePath));
  zip.writeZip(outputPath);
  const elapsed = ((performance.now() - startTime) / 1000).toFixed(3);
  console.log(`Created backup: ${filename} (${elapsed}s)`);
}

async function rotateBackups(backupDir) {
  log('[DEBUG] Starting rotateBackups');
  const now = new Date();
  const files = await fs.readdir(backupDir);
  log('[DEBUG] Files read:', files.length);
  const backupFiles = files.filter(f => f.startsWith('export') && f.endsWith('.zip'));
  log('[DEBUG] Backup files found:', backupFiles.length);

  const backups = backupFiles.map(f => {
    const dateStr = f.slice(6, -4); // exportYYYYMMdd.zip -> YYYYMMdd
    const date = parse(dateStr);
    return { file: f, date, year: date.getFullYear() };
  }).filter(b => !isNaN(b.date.getTime()));

  const toKeep = new Set();

  // Keep daily for last 7 days (today - 6 days)
  log('[DEBUG] Processing daily backups');
  for (let i = 0; i < 7; i++) {
    const d = subDays(now, i);
    const key = `export${format(d)}.zip`;
    toKeep.add(key);
  }
  log('[DEBUG] Daily backups processed');

  // Keep weekly (end of week: Sunday) for earlier in current month
  log('[DEBUG] Processing weekly backups');
  const currentMonthStart = startOfMonth(now);
  let weekEndDate = endOfWeek(subDays(now, 7));
  let weekCount = 0;
  while (isSameMonth(weekEndDate, now) && weekEndDate >= currentMonthStart && weekCount < 10) {
    const key = `export${format(weekEndDate)}.zip`;
    toKeep.add(key);
    weekEndDate = endOfWeek(subDays(weekEndDate, 7));
    weekCount++;
  }
  log('[DEBUG] Weekly backups processed');

  // Keep monthly (last day of month) for previous months in current year
  log('[DEBUG] Processing monthly backups');
  for (let m = 1; m < now.getMonth() + 1; m++) {
    const prevMonthEnd = endOfMonth(new Date(now.getFullYear(), now.getMonth() - m, 1));
    const key = `export${format(prevMonthEnd)}.zip`;
    toKeep.add(key);
  }
  log('[DEBUG] Monthly backups processed');

  // Keep yearly (Dec 31) for previous years, based on existing backups
  log('[DEBUG] Processing yearly backups');
  if (backups.length > 0) {
    const minYear = Math.min(...backups.map(b => b.year));
    for (let y = minYear; y < now.getFullYear(); y++) {
      const yearEnd = endOfYear(new Date(y, 0, 1));
      const key = `export${format(yearEnd)}.zip`;
      toKeep.add(key);
    }
  }
  log('[DEBUG] Yearly backups processed');

  // Delete files not in toKeep
  log('[DEBUG] Processing deletions');
  for (const { file } of backups) {
    if (!toKeep.has(file)) {
      const filePath = path.join(backupDir, file);
      await fs.unlink(filePath);
      console.log(`Deleted old backup: ${file}`);
    }
  }
  log('[DEBUG] rotateBackups completed');
}

async function main() {
  const filePathsArg = process.argv[2];
  const backupDir = process.argv[3];

  if (!filePathsArg || !backupDir) {
    console.error('Usage: node ark.js <filePath1[,filePath2,...]> <backupDir>');
    console.error('Example: node ark.js /path/to/xy.db /opt/backup');
    console.error('Example: node ark.js /path/to/xy.db,/path/to/xy.sql /opt/backup');
    process.exit(1);
  }

  const filePaths = filePathsArg.split(',').map(f => f.trim());

  try {
    await fs.mkdir(backupDir, { recursive: true });
    await createBackup(filePaths, backupDir);
    await rotateBackups(backupDir);
    console.log('Backup and rotation completed.');
    process.exit(0);
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();
